tests/template/test_clust_template.py

```diff
@@ -0,0 +1,75 @@
+import io
+import os
+import numpy as np
+import nbformat
+from temp_utils import _get_hdf5_params
+from libertem.analysis.clust import ClusterAnalysis
+from libertem.web.notebook_generator.notebook_generator import notebook_generator
+from nbconvert.preprocessors import ExecutePreprocessor
+from libertem.udf.stddev import consolidate_result
+
+
+class ClusterAnalysisTest(ClusterAnalysis):
+    def run_sd_udf(self, roi, stddev_udf, *args):
+        sd_udf_results = self.ctx.run_udf(
+                            dataset=self.dataset,
+                            udf=stddev_udf,
+                            roi=roi,
+                            progress=True
+        )
+        return roi, consolidate_result(sd_udf_results)
+
+    def controller(self, ctx):
+        self.ctx = ctx
+        roi, sd_udf_results = self.get_sd_results(None, None, None)
+        cluster_udf = self.get_cluster_udf(sd_udf_results)
+        udf_results = ctx.run_udf(dataset=self.dataset, udf=cluster_udf)
+        cluster_result = self.get_udf_results(udf_results, roi)
+        return cluster_result
+
+
+def test_clust_default(hdf5_ds_2, tmpdir_factory, inline_executor, lt_ctx):
+    datadir = tmpdir_factory.mktemp('template_tests')
+
+    conn = {'connection': {'type': 'local'}}
+    path = hdf5_ds_2.path
+    dataset = _get_hdf5_params(path)
+
+    analysis = [{
+            "analysisType": 'CLUST',
+            "parameters": {
+                    'n_peaks': 42,
+                    'n_clust': 7,
+                    'cy': 3,
+                    'cx': 3,
+                    'ri': 1,
+                    'ro': 5,
+                    'delta': 0.05,
+                    'min_dist': 1,
+                    }
+    }]
+
+    notebook = notebook_generator(conn, dataset, analysis, save=True)
+    notebook = io.StringIO(notebook.getvalue())
+    nb = nbformat.read(notebook, as_version=4)
+    ep = ExecutePreprocessor(timeout=600)
+    out = ep.preprocess(nb, {"metadata": {"path": datadir}})
+    data_path = os.path.join(datadir, 'cluster_result.npy')
+    results = np.load(data_path)
+
+    analysis = ClusterAnalysisTest(dataset=hdf5_ds_2, parameters={
```

Comment:
I would prefer if this tested against the "real" `ClusterAnalysis`. I guess the challenge was the `async` methods? It should be possible to use a `Context` with an `async_executor` instead, mark the test as an `asyncio` test and use the original class here. For "faking" `send_results`, you could do something like this:

```python
class ResultContainer:
    def __call__(self, results, finished):
        self.results = results
        self.finished = finished

res_container = ResultContainer()
# ...
await analysis.controller(..., send_results=res_container)
```

Then you can grab the results from `res_container.results`. Or you can use a `unittest.mock.Mock` object and assert on the `call_args_list`. What do you think?

Reply:
Sounds good! I was a little confused earlier.